File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/Observer/concrete_subject.go
package Observer

import "sync"

// Subject 具体主题
type Subject struct {
	name     string
	password string
	changeMgr *SimpleChangeManager
}

var subjectInstance *Subject
var subjectOnce sync.Once

// GetSubject 获取主题实例
func GetSubject() *Subject {
	subjectOnce.Do(func() {
		subjectInstance = &Subject{
			changeMgr: NewSimpleChangeManager(subjectInstance),
		}
		// 重新设置changeMgr的subject为当前实例
		subjectInstance.changeMgr = NewSimpleChangeManager(subjectInstance)
	})
	return subjectInstance
}

// RegisterAccount 注册账户
func (s *Subject) RegisterAccount(name, password string) {
	s.name = name
	s.password = password
	s.Attach(GetRegisterObserver())
}

// LoginAccount 登录账户
func (s *Subject) LoginAccount(name, password string) {
	s.name = name
	s.password = password
	s.Attach(GetLoginObserver())
}

// GetState 获取状态
func (s *Subject) GetState(state int) string {
	switch state {
	case 0:
		return s.name
	case 1:
		return s.password
	default:
		return ""
	}
}

// Attach 附加观察者
func (s *Subject) Attach(observer IObserver) {
	s.changeMgr.Register(s, observer)
}

// Detach 分离观察者
func (s *Subject) Detach(observer IObserver) {
	s.changeMgr.Unregister(s, observer)
}

// Notify 通知观察者
func (s *Subject) Notify() {
	s.changeMgr.Notify()
	s.changeMgr.ClearObserver()
}

File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/Observer/ConcreteSubject.lua
local ObserverModule = require("Observer")
local ISubject = ObserverModule.ISubject
local ChangeManagerModule = require("ChangeManager")
local SimpleChangeManager = ChangeManagerModule.SimpleChangeManager

-- 具体主题
local Subject = {}
Subject.__index = Subject
setmetatable(Subject, ISubject)

local subjectInstance = nil

function Subject:getInstance()
    if not subjectInstance then
        subjectInstance = Subject:new()
    end
    return subjectInstance
end

function Subject:new()
    local o = ISubject:new()
    setmetatable(o, self)
    o.name = ""
    o.password = ""
    o.changeMgr = SimpleChangeManager:new(o)
    return o
end

function Subject:RegisterAccount(name, password)
    self.name = name
    self.password = password
    self:Attach(require("ConcreteObserver").getRegisterObserver())
end

function Subject:LoginAccount(name, password)
    self.name = name
    self.password = password
    self:Attach(require("ConcreteObserver").getLoginObserver())
end

function Subject:GetState(state)
    if state == 0 then
        return self.name
    elseif state == 1 then
        return self.password
    else
        return ""
    end
end

function Subject:Attach(observer)
    self.changeMgr:Register(self, observer)
end

function Subject:Detach(observer)
    self.changeMgr:Unregister(self, observer)
end

function Subject:Notify()
    self.changeMgr:Notify()
    self.changeMgr:ClearObserver()
end

return Subject
